x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/AdjustableCapacityBlockingQueue.java

```diff
@@ -7,66 +7,86 @@
+     * @param initialCapacity the initial capacity of the queue, if null the queue will be unbounded
      */
-    public AdjustableCapacityBlockingQueue(int initialCapacity, Function<Integer, BlockingQueue<E>> createQueue) {
-        this.createQueue = Objects.requireNonNull(createQueue);
-        queue = createQueue.apply(initialCapacity);
+    public AdjustableCapacityBlockingQueue(QueueCreator<E> queueCreator, @Nullable Integer initialCapacity) {
+        this.queueCreator = Objects.requireNonNull(queueCreator);
+        currentQueue = createCurrentQueue(queueCreator, initialCapacity);
         lock = new ReentrantReadWriteLock();
+        prioritizedReadingQueue = queueCreator.create();
+    }
+
+    private static <E> BlockingQueue<E> createCurrentQueue(QueueCreator<E> queueCreator, @Nullable Integer initialCapacity) {
+        if (initialCapacity == null) {
+            return queueCreator.create();
+        }
+
+        return queueCreator.create(initialCapacity);
     }
 
     /**
      * Sets the capacity of the queue. If the new capacity is smaller than the current number of elements in the queue, the
-     * elements that exceed the new capacity size are returned and will not be present in the queue.
-     * This is potentially an expensive operation because a new internal queue is instantiated.
-     * @param capacity the new capacity for the queue.
-     * @return a list of elements that could not fit within the queue given the new capacity requirement.
+     * elements that exceed the new capacity are retained. In this situation the {@link AdjustableCapacityBlockingQueue#size()} method
+     * could return a value greater than the specified capacity.
+     * <br/>
+     * <b>This is potentially an expensive operation because a new internal queue is instantiated.</b>
+     * @param capacity the new capacity for the queue
      */
-    public List<E> setCapacity(int capacity) {
+    public void setCapacity(int capacity) {
         final ReentrantReadWriteLock.WriteLock writeLock = lock.writeLock();
-        List<E> remainingItems = new ArrayList<>(queue.size());
 
         writeLock.lock();
         try {
-            BlockingQueue<E> newQueue = createQueue.apply(capacity);
-            queue.drainTo(newQueue, capacity);
-            queue.drainTo(remainingItems);
-            queue = newQueue;
+            BlockingQueue<E> newQueue = queueCreator.create(capacity);
+            // Drain the first items from the queue, so they will get read first.
+            // Only drain the amount that wouldn't fit in the new queue
+            // If the new capacity is larger than the current queue size then we don't need to drain any
+            // they will all fit within the newly created queue. In this situation the queue size - capacity
+            // would result in a negative value which is ignored
+            if (currentQueue.size() - capacity > 0) {
```

Comment:
nit: I think if we had `if (currentQueue.size()  > capacity)` it would be easier to understand what is happening here.